common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java

```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
     }
   }
 
+  private class ManagedBufferIterator implements Iterator<ManagedBuffer> {
+
+    private int index = 0;
+    private String appId;
+    private String execId;
+    private String shuffleId;
+    // An array containing mapId and reduceId pairs.
+    private int[] mapIdAndReduceIds;
+
+    ManagedBufferIterator(String appId, String execId, String[] blockIds) {
+      this.appId = appId;
+      this.execId = execId;
+      String[] blockId0Parts = blockIds[0].split("_");
+      if (blockId0Parts.length < 4) {
+        throw new IllegalArgumentException("Unexpected block id format: " + blockIds[0]);
+      } else if (!blockId0Parts[0].equals("shuffle")) {
+        throw new IllegalArgumentException("Expected shuffle block id, got: " + blockIds[0]);
+      }
+      this.shuffleId = blockId0Parts[1];
+      mapIdAndReduceIds = new int[2 * blockIds.length];
+      for (int i = 0; i< blockIds.length; i++) {
+        String[] blockIdParts = blockIds[i].split("_");
+        if (!blockIdParts[1].equals(shuffleId)) {
+          throw new IllegalArgumentException("Expected shuffleId=" + shuffleId +
+            ", got:" + blockIds[i]);
+        }
+        mapIdAndReduceIds[2 * i] = Integer.parseInt(blockIdParts[2]);
+        mapIdAndReduceIds[2 * i + 1] = Integer.parseInt(blockIdParts[3]);
+      }
+    }
+
+    @Override
+    public boolean hasNext() {
+      return index < mapIdAndReduceIds.length / 2;
+    }
+
+    @Override
+    public ManagedBuffer next() {
+      String blockId = "shuffle_" + shuffleId + "_" + mapIdAndReduceIds[2 * index] + "_" +
+        mapIdAndReduceIds[2 * index + 1];
+      final ManagedBuffer block = blockManager.getBlockData(appId, execId, blockId);
```

Comment:
Is it too big a change to make `getBlockData` expect the broken down shuffle/map/reduce id, since all it does is parse the string again into integers?